src/hotspot/share/opto/compile.cpp

```diff
@@ -2188,11 +2190,17 @@ void Compile::Optimize() {
       if (failing())  return;
 
       if (congraph() != NULL && macro_count() > 0) {
+#ifndef PRODUCT
+        int _prev_scalar_replaced = Atomic::load(&PhaseMacroExpand::_objs_scalar_replaced_counter);
+#endif
         TracePhase tp("macroEliminate", &timers[_t_macroEliminate]);
         PhaseMacroExpand mexp(igvn);
         mexp.eliminate_macro_nodes();
         igvn.set_delay_transform(false);
 
+#ifndef PRODUCT
+        congraph()->update_escape_state(Atomic::load(&PhaseMacroExpand::_objs_scalar_replaced_counter) - _prev_scalar_replaced);
```

Comment:
I understand that you would like to add scalarized objects back to the snapshot, but this _objs_scalar_replaced_counter is a static counter as well. Two consecutive atomic loads are not helpful here because other C2 compiler threads can update it.

I think we can a member data ConnectionGraph::_prev_scalar_replaced and increment it in mexp.eliminate_macro_nodes();  we keep _objs_scalar_replaced_counter  but only atomic accumulate it at the end of ME phrase.

Reply:
@navyxliu I followed the general idea of your advice and created _local_scalar_replaced, which gets added in compiler.cpp.